<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>1&nbsp; Preparing DNA data for training – Biological language models &amp; Neural Networks</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Chapter2.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-687258f6a899824ebad24e93c100d594.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Biological language models &amp; Neural Networks</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Chapter1.html">DNA</a></li><li class="breadcrumb-item"><a href="./Chapter1.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Preparing DNA data for training</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">DNA</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter1.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Preparing DNA data for training</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Training our first DNA Language Model</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Evaluating DNA Language Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Chapter4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Chapter 4 - Evolution-Aware BERT: Designing Encoders for Genomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#garbage-in-garbage-out" id="toc-garbage-in-garbage-out" class="nav-link active" data-scroll-target="#garbage-in-garbage-out"><span class="header-section-number">1.1</span> Garbage in garbage out</a></li>
  <li><a href="#understanding-ensembl-and-biomart" id="toc-understanding-ensembl-and-biomart" class="nav-link" data-scroll-target="#understanding-ensembl-and-biomart"><span class="header-section-number">1.2</span> Understanding Ensembl and BioMart</a>
  <ul class="collapse">
  <li><a href="#what-are-fasta-files" id="toc-what-are-fasta-files" class="nav-link" data-scroll-target="#what-are-fasta-files"><span class="header-section-number">1.2.1</span> What Are FASTA Files?</a></li>
  <li><a href="#why-focus-on-coding-dna-sequences-cds" id="toc-why-focus-on-coding-dna-sequences-cds" class="nav-link" data-scroll-target="#why-focus-on-coding-dna-sequences-cds"><span class="header-section-number">1.2.2</span> Why Focus on Coding DNA Sequences (CDS)?</a></li>
  </ul></li>
  <li><a href="#why-upload-dna-data-to-hugging-face" id="toc-why-upload-dna-data-to-hugging-face" class="nav-link" data-scroll-target="#why-upload-dna-data-to-hugging-face"><span class="header-section-number">1.3</span> Why Upload DNA Data to Hugging Face?</a></li>
  <li><a href="#the-script-downloading-and-formatting-human-cds-data" id="toc-the-script-downloading-and-formatting-human-cds-data" class="nav-link" data-scroll-target="#the-script-downloading-and-formatting-human-cds-data"><span class="header-section-number">1.4</span> The Script: Downloading and Formatting Human CDS Data</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">1.5</span> Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Chapter1.html">DNA</a></li><li class="breadcrumb-item"><a href="./Chapter1.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Preparing DNA data for training</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Preparing DNA data for training</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Abstract
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this chapter we’ll learn how to create a high fidelity DNA datasets. Two datasets I have created are available on Huggingface: <a href="https://huggingface.co/datasets/MichelNivard/Human-genome-CDS-GRCh38">the human CDS</a> and <a href="https://huggingface.co/datasets/MichelNivard/Coding_Sequences_13_Species">the CDS of 13 vertebrae species</a>.</p>
<p>Scripts for this chapter are found here: <a href="https://github.com/MichelNivard/Biological-language-models/tree/main/scripts/Chapter_1" class="uri">https://github.com/MichelNivard/Biological-language-models/tree/main/scripts/Chapter_1</a></p>
</div>
</div>
<section id="garbage-in-garbage-out" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="garbage-in-garbage-out"><span class="header-section-number">1.1</span> Garbage in garbage out</h2>
<p>High-quality data is essential for training effective machine learning models. In natural language processing (NLP), researchers invest significant effort in curating clean datasets from raw internet text. For example a dataset like ‘fineweb-edu’ contains English text that is of very high quality. Models trained on fineweb-edu (and similar high quality datasets) will improve MUCH faster then the equivalent model trained on other less carefully processed and evaluated datasets.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://cdn-uploads.huggingface.co/production/uploads/61c141342aac764ce1654e43/QqXOM8h_ZjjhuCv71xmV7.png" class="img-fluid figure-img"></p>
<figcaption>Relative training efficiency using a high quality dataset like fineweb-edu and other massive datasets of English language texts. Image obtained from: https://huggingface.co/datasets/HuggingFaceFW/fineweb-edu</figcaption>
</figure>
</div>
<p>Those with experience with <strong>genetics</strong> will know most of what I am about to explain when it comes to the sources of genetic data and the file formats these come in. Those with an <strong>ML</strong> background will be very familiar with Huggingface and the amazing data/model/training integration their libraries offer. Assembing high quality genomics datasets for language modeling requires familiarity with both. When working with DNA, RNA, and protein sequences, the authoritative sources of data are specialized databases and infrastructures rather than data scraped from the internet. If you want to learn to train DNA/RNA/Protein-based language model, we must learn how, and where, to retrieve data and convert it into a structured format.</p>
<p>In this tutorial, we will begin by collecting coding DNA sequences (CDS) from the human genome using the Ensembl database via BioMart. The goal is to store these sequences in a format suitable for training a DNA-based language model. We will then upload the processed data to <a href="https://huggingface.co">Huggingface</a>, a platform for hosting datasets and models for machine learning tasks.</p>
<p>This chapter will introduce key biological and computational concepts, ensuring that both biology newcomers and those unfamiliar with language modeling tools can follow along.</p>
</section>
<section id="understanding-ensembl-and-biomart" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="understanding-ensembl-and-biomart"><span class="header-section-number">1.2</span> Understanding Ensembl and BioMart</h2>
<p>Fortunately for us, there is decades of work cleaning up genomic data and we can just go and get it from US government funded websites, where it is deposited by the global scientific community. <strong>Ensembl</strong> is a genomic database that provides extensive annotations for various species, including humans. It offers access to gene sequences, transcript annotations, and protein-coding information. One of its most powerful tools is <strong>BioMart</strong>, a flexible data retrieval system that allows users to download specific genomic datasets easily.</p>
<p>If we want t work with the data in a language model its efficient to store it in a format that is tailored for machine learning libraries. Unlike text-based NLP datasets, genomic data is structured differently and must be properly formatted before use in a model.</p>
<section id="what-are-fasta-files" class="level3" data-number="1.2.1">
<h3 data-number="1.2.1" class="anchored" data-anchor-id="what-are-fasta-files"><span class="header-section-number">1.2.1</span> What Are FASTA Files?</h3>
<p>A <strong>FASTA file</strong> is a simple text-based format used for storing biological sequences. Each entry in a FASTA file consists of: 1. A <strong>header line</strong> (starting with <code>&gt;</code>), which contains metadata such as gene IDs and chromosome locations. 2. A <strong>sequence</strong> line, which contains the nucleotide or protein sequence.</p>
<p>There is a very comprehensive Wikipedia entry on the FASTA format.</p>
<p>“Sequences may be&nbsp;protein sequences&nbsp;or&nbsp;nucleic acid&nbsp;sequences, and they can contain gaps or alignment characters (see&nbsp;sequence alignment). Sequences are expected to be represented in the standard IUB/IUPAC&nbsp;amino acid&nbsp;and&nbsp;nucleic acid&nbsp;codes, with these exceptions: lower-case letters are accepted and are mapped into upper-case; a single hyphen or dash can be used to represent a gap character; and in amino acid sequences, U and * are acceptable letters (see below). Numerical digits are not allowed but are used in some databases to indicate the position in the sequence.” ((source: <a href="https://en.wikipedia.org/wiki/FASTA_format" class="uri">https://en.wikipedia.org/wiki/FASTA_format</a>))</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;">Nucleic Acid Code</th>
<th style="text-align: center;">Meaning</th>
<th style="text-align: center;">Mnemonic</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Adenine" title="Adenine"><strong>A</strong>denine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">C</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Cytosine" title="Cytosine"><strong>C</strong>ytosine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">G</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Guanine" title="Guanine"><strong>G</strong>uanine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Thymine" title="Thymine"><strong>T</strong>hymine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">U</td>
<td style="text-align: center;">U</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Uracil" title="Uracil"><strong>U</strong>racil</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">(i)</td>
<td style="text-align: center;">i</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Inosine" title="Inosine"><strong>i</strong>nosine</a>&nbsp;(non-standard)</td>
</tr>
<tr class="odd">
<td style="text-align: center;">R</td>
<td style="text-align: center;">A or G (I)</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Purine" title="Purine">pu<strong>R</strong>ine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">Y</td>
<td style="text-align: center;">C, T or U</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Pyrimidine" title="Pyrimidine">p<strong>Y</strong>rimidines</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">K</td>
<td style="text-align: center;">G, T or U</td>
<td style="text-align: center;">bases which are&nbsp;<a href="https://en.wikipedia.org/wiki/Ketone" title="Ketone"><strong>K</strong>etones</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">M</td>
<td style="text-align: center;">A or C</td>
<td style="text-align: center;">bases with&nbsp;<a href="https://en.wikipedia.org/wiki/Amino" title="Amino">a<strong>M</strong>ino groups</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">S</td>
<td style="text-align: center;">C or G</td>
<td style="text-align: center;"><strong>S</strong>trong interaction</td>
</tr>
<tr class="even">
<td style="text-align: center;">W</td>
<td style="text-align: center;">A, T or U</td>
<td style="text-align: center;"><strong>W</strong>eak interaction</td>
</tr>
<tr class="odd">
<td style="text-align: center;">B</td>
<td style="text-align: center;">not A (i.e.&nbsp;C, G, T or U)</td>
<td style="text-align: center;"><strong>B</strong>&nbsp;comes after A</td>
</tr>
<tr class="even">
<td style="text-align: center;">D</td>
<td style="text-align: center;">not C (i.e.&nbsp;A, G, T or U)</td>
<td style="text-align: center;"><strong>D</strong>&nbsp;comes after C</td>
</tr>
<tr class="odd">
<td style="text-align: center;">H</td>
<td style="text-align: center;">not G (i.e., A, C, T or U)</td>
<td style="text-align: center;"><strong>H</strong>&nbsp;comes after G</td>
</tr>
<tr class="even">
<td style="text-align: center;">V</td>
<td style="text-align: center;">neither T nor U (i.e.&nbsp;A, C or G)</td>
<td style="text-align: center;"><strong>V</strong>&nbsp;comes after U</td>
</tr>
<tr class="odd">
<td style="text-align: center;">N</td>
<td style="text-align: center;">A C G T U</td>
<td style="text-align: center;"><strong>N</strong>ucleic acid</td>
</tr>
<tr class="even">
<td style="text-align: center;">-</td>
<td style="text-align: center;">gap of indeterminate length</td>
<td style="text-align: center;"></td>
</tr>
</tbody>
</table>
<p>The amino acid codes supported (22 amino acids and 3 special codes) are:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;">Amino Acid Code</th>
<th style="text-align: center;">Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">A</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Alanine" title="Alanine">Alanine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">B</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Aspartic_acid" title="Aspartic acid">Aspartic acid</a>&nbsp;(D) or&nbsp;<a href="https://en.wikipedia.org/wiki/Asparagine" title="Asparagine">Asparagine</a>&nbsp;(N)</td>
</tr>
<tr class="odd">
<td style="text-align: center;">C</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Cysteine" title="Cysteine">Cysteine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">D</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Aspartic_acid" title="Aspartic acid">Aspartic acid</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">E</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Glutamic_acid" title="Glutamic acid">Glutamic acid</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">F</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Phenylalanine" title="Phenylalanine">Phenylalanine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">G</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Glycine" title="Glycine">Glycine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">H</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Histidine" title="Histidine">Histidine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">I</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Isoleucine" title="Isoleucine">Isoleucine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">J</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Leucine" title="Leucine">Leucine</a>&nbsp;(L) or&nbsp;<a href="https://en.wikipedia.org/wiki/Isoleucine" title="Isoleucine">Isoleucine</a>&nbsp;(I)</td>
</tr>
<tr class="odd">
<td style="text-align: center;">K</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Lysine" title="Lysine">Lysine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">L</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Leucine" title="Leucine">Leucine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">M</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Methionine" title="Methionine">Methionine</a>/<a href="https://en.wikipedia.org/wiki/Start_codon" title="Start codon">Start codon</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">N</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Asparagine" title="Asparagine">Asparagine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">O</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Pyrrolysine" title="Pyrrolysine">Pyrrolysine</a>&nbsp;(rare)</td>
</tr>
<tr class="even">
<td style="text-align: center;">P</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Proline" title="Proline">Proline</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">Q</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Glutamine" title="Glutamine">Glutamine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">R</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Arginine" title="Arginine">Arginine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">S</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Serine" title="Serine">Serine</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Threonine" title="Threonine">Threonine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">U</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Selenocysteine" title="Selenocysteine">Selenocysteine</a>&nbsp;(rare)</td>
</tr>
<tr class="even">
<td style="text-align: center;">V</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Valine" title="Valine">Valine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">W</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Tryptophan" title="Tryptophan">Tryptophan</a></td>
</tr>
<tr class="even">
<td style="text-align: center;">Y</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Tyrosine" title="Tyrosine">Tyrosine</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">Z</td>
<td style="text-align: center;"><a href="https://en.wikipedia.org/wiki/Glutamic_acid" title="Glutamic acid">Glutamic acid</a>&nbsp;(E) or&nbsp;<a href="https://en.wikipedia.org/wiki/Glutamine" title="Glutamine">Glutamine</a>&nbsp;(Q)</td>
</tr>
<tr class="even">
<td style="text-align: center;">X</td>
<td style="text-align: center;">any</td>
</tr>
<tr class="odd">
<td style="text-align: center;">*</td>
<td style="text-align: center;">translation stop</td>
</tr>
<tr class="even">
<td style="text-align: center;">-</td>
<td style="text-align: center;">gap of indeterminate length</td>
</tr>
</tbody>
</table>
<p>You’ll notice the FASTA format has a well defined structure, and it could be leveraged to build a complete tokenizer, for now though our 4 character (+6 special characters) tokenizer will have to do.</p>
</section>
<section id="why-focus-on-coding-dna-sequences-cds" class="level3" data-number="1.2.2">
<h3 data-number="1.2.2" class="anchored" data-anchor-id="why-focus-on-coding-dna-sequences-cds"><span class="header-section-number">1.2.2</span> Why Focus on Coding DNA Sequences (CDS)?</h3>
<p>In the example, we retrieve the <strong>human coding DNA sequences (CDS)</strong>, which represent the DNA sequence of protein-coding regions of genes.</p>
<p>While our ultimate goal is to model the entire human genome—and potentially multiple genomes across species or individuals—such tasks require significant computational resources. The tutorials in this book are designed to be accessible, running efficiently on a MacBook or Google Colab. Therefore, we focus on <strong>CDS</strong>, which are highly structured DNA sequences within genes, that directly transcribed into RNA which is in turn translate into proteins. the Table below contains the direct translation from 3 letter DNA sequence to amino-acid (which are the building blocks of proteins).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/paste-3.png" class="img-fluid figure-img"></p>
<figcaption>The Genetic code to translate codins (3 leter DNA sequences) to amino-acids that are in turn the building blocks of proteins (source: https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/_images/P7_image1.png)</figcaption>
</figure>
</div>
<p>In contrast, much of the rest of the genome consists of regulatory regions, which are more complex and less structured. CDS sequences provide a strong foundation for a small DNA-based language model because they contain well-defined biological structure, making them a logical starting point before expanding to more complex genomic elements.</p>
</section>
</section>
<section id="why-upload-dna-data-to-hugging-face" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="why-upload-dna-data-to-hugging-face"><span class="header-section-number">1.3</span> Why Upload DNA Data to Hugging Face?</h2>
<p>Hugging Face provides a robust ecosystem for hosting and sharing datasets, particularly for machine learning applications. Some key advantages include: - <strong>Easy accessibility</strong>: Researchers and models can easily retrieve datasets. - <strong>Standardized format</strong>: Datasets are structured for seamless integration with deep learning frameworks. - <strong>Direct integration with Hugging Face tools</strong>: The data on the Hugging Face Hub integrates seamlessly with their <strong>Transformers</strong> and <strong>Trainer</strong> Python libraries, making it easy to load datasets and train models. - <strong>Version control and updates</strong>: Data can be refined and expanded over time.</p>
<p>By storing our dataset on Hugging Face, we enable efficient training and collaboration for DNA language modeling.</p>
</section>
<section id="the-script-downloading-and-formatting-human-cds-data" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="the-script-downloading-and-formatting-human-cds-data"><span class="header-section-number">1.4</span> The Script: Downloading and Formatting Human CDS Data</h2>
<p>Below is the R script that downloads human CDS from Ensembl using BioMart, extracts metadata, and saves the processed data into a CSV file. the package we use, <code>biomartr</code> isn’t the official R package but its a great option! it has very extensive documentation, so if you want to download other sequences in the future make sure to start here: <a href="https://docs.ropensci.org/biomartr/" class="uri">https://docs.ropensci.org/biomartr/</a></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install necessary packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"biomartr"</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required libraries</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomartr)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the human CDS dataset from Ensembl</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>HS.cds.ensembl <span class="ot">&lt;-</span> <span class="fu">getCDS</span>(<span class="at">db =</span> <span class="st">"ensembl"</span>, <span class="at">organism =</span> <span class="st">"Homo sapiens"</span>, <span class="at">path =</span> <span class="fu">file.path</span>(<span class="st">"_ncbi_downloads"</span>,<span class="st">"CDS"</span>))</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Read CDS data as a Biostrings object</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>Human_CDS <span class="ot">&lt;-</span> <span class="fu">read_cds</span>(<span class="at">file =</span> HS.cds.ensembl, <span class="at">obj.type =</span> <span class="st">"Biostrings"</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract headers and sequences</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>headers <span class="ot">&lt;-</span> <span class="fu">names</span>(Human_CDS)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">as.character</span>(Human_CDS)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to extract metadata from headers</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>extract_metadata <span class="ot">&lt;-</span> <span class="cf">function</span>(header) {</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>  transcript_id <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"^&gt;([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  chromosome <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*chromosome:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  start <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*chromosome:[^:]+:([^:]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  end <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*chromosome:[^:]+:[^:]+:([^:]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  strand <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*chromosome:[^:]+:[^:]+:[^:]+:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  gene_id <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*gene:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  gene_biotype <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*gene_biotype:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  transcript_biotype <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*transcript_biotype:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  gene_symbol <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*gene_symbol:([^ ]+).*"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>  description <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*description:(.*)"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, header)</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">transcript_id =</span> transcript_id,</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">chromosome =</span> chromosome,</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">start =</span> start,</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">end =</span> end,</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">strand =</span> strand,</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene_id =</span> gene_id,</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene_biotype =</span> gene_biotype,</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">transcript_biotype =</span> transcript_biotype,</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene_symbol =</span> gene_symbol,</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">description =</span> description</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract metadata from headers</span></span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>metadata_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(headers, extract_metadata)</span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>metadata_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(metadata_list, as.data.frame))</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>metadata_df<span class="sc">$</span>sequence <span class="ot">&lt;-</span> sequences</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Save data to CSV</span></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(metadata_df, <span class="st">"genome_sequences.csv"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Print sample data</span></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can run the script yourself, but I have also gone ahead and uploaded it to huggingface: <a href="https://huggingface.co/datasets/MichelNivard/Human-genome-CDS-GRCh38" class="uri">https://huggingface.co/datasets/MichelNivard/Human-genome-CDS-GRCh38</a></p>
</section>
<section id="summary" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="summary"><span class="header-section-number">1.5</span> Summary</h2>
<p>In this chapter, we: - Introduced <strong>Ensembl</strong> and <strong>BioMart</strong> as tools for retrieving genomic data. - Explained <strong>FASTA files</strong> and <strong>human CDS</strong>, which form the core of our dataset. - Discussed the advantages of <strong>uploading datasets to Hugging Face</strong>, emphasizing its integration with <strong>Transformers</strong> and <strong>Trainer</strong> libraries. - Provided an <strong>R script</strong> to download, process, and store human CDS in a structured format.</p>
<p>In the next chapter, we will explore <strong>preprocessing techniques like tokenization</strong> and strategies for encoding DNA sequences into a format suitable for training a deep learning language model, and we use Huggingface Transformers and Trainer library to train our first little DNA language model!</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Preface">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Chapter2.html" class="pagination-link" aria-label="Training our first DNA Language Model">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Training our first DNA Language Model</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>